### PR TITLE
Enable torch.autocast with ZeRO

### PR DESCRIPTION
**Pull request by tohtana**
This PR supports `torch.autocast` with ZeRO. You need to enable `torch.autocast` in DeepSpeed config.

```json
"torch_autocast": {
  "enabled": true,
  "dtype": "bfloat16"
}
```

You don't need to explicitly call `torch.autocast` in your code. The grad scaler is also applied in the DeepSpeed optimizer.

All the parameters are maintained in FP32 but certain operators and layers are computed in the specified dtype. With ZeRO enabled, the communication (allreduce/reduce_scatter) for specified layers are also done in the specified dtype ([List of modules](https://github.com/microsoft/DeepSpeed/blob/0e52edb6a3cc940a9d8c53f7343f3473ebd69bbb/deepspeed/runtime/torch_autocast.py)).
You cannot enable `fp16` or `bf16` with `torch.autocast`.

(Currently working on ZeRO3)

### CODE REVIEW

**Comment by stas00** [comment]
I have a question about `lower_precision_safe_modules`

https://pytorch.org/docs/stable/amp.html#torch.autocast doesn't have an option to specify `lower_precision_safe_modules` - why do we then put the onus on the deepspeed user? (and I'm aware that it's optional and there is a default list). 

My question is - can we automatically retrieve that safe list from pytorch? Or is this because what's pytorch considers safe isn't necessarily what deepspeed considers safe?

File: deepspeed/runtime/constants.py [hunk 1]
```diff
@@ -202,6 +202,23 @@
...
 
+#########################################
+# Torch AMP support
+#########################################
+TORCH_AUTOCAST_FORMAT = '''
+PyTorch autocast config should be of the format:
+"torch_autocast": {
+  "enabled": true,
+  "dtype": "bfloat16",
+}
```
Comment (stas00) [inline deepspeed/runtime/constants.py:213]: add `...` here as in other sections as it's incomplete.

Comment (stas00) [reply deepspeed/runtime/constants.py:213]: or update it with the rest of flags?

**Review by stas00 — COMMENTED** [review]
If the training loop uses `torch.autocast` 
1. could deepspeed detect that and probably assert if ds config isn't done right?
2. or actually should it assert if this happens because then the behavior is unsupported?

**Comment by tohtana** [comment]
Hi @stas00, sorry I have missed your replies.

> My question is - can we automatically retrieve that safe list from pytorch? Or is this because what's pytorch considers safe isn't necessarily what deepspeed considers safe?

Great question. Ideally, we would retrieve the list directly from PyTorch. I found it in the following files:
- https://github.com/pytorch/pytorch/blob/18a7a04c4adecda3be17dd364d48d484fd1dcdba/aten/src/ATen/autocast_mode.cpp
- https://github.com/pytorch/pytorch/blob/18a7a04c4adecda3be17dd364d48d484fd1dcdba/aten/src/ATen/autocast_mode.h#L828

It does seem challenging to determine which module parameters are safe to use in lower precision. Hopefully, we can refine the default list for this feature based on real-world usage.

**Comment by tohtana** [comment]
> 1. could deepspeed detect that and probably assert if ds config isn't done right?
> 2. or actually should it assert if this happens because then the behavior is unsupported?

I think it should be harmless to include `torch.autocast` in the training loop, though we should display a warning message. Let me double-check the behavior and add that message.

One confusing aspect is that it's valid to use PyTorch’s `torch.autocast` while disabling "torch_autocast" in the DeepSpeed config. In this case, DeepSpeed treats it as a pure FP32 setting. The main issue is that all communication operations (e.g., gradient reduction and all-gather) will also be performed in FP32. Since some modules are computed in BF16/FP16, this leads to inefficiencies in many cases.

**Comment by stas00** [comment]
> One confusing aspect is that it's valid to use PyTorch’s torch.autocast while disabling "torch_autocast" in the DeepSpeed config. In this case, DeepSpeed treats it as a pure FP32 setting.

Dare I say this is not valid - nobody does fp32 training these days.

Such situations should assert IMHO.

And fp32 comms is a huge problem as well, not just because of time overhead, but because the comms buffer is 2x larger and this happens 2 times - so overall it'll consume 4x more memory for all comms. I happen to know this because I discovered this when porting Ulysses SP, where it automatically enables fp32 comms and I had huge memory spikes in backward because of that.

**Comment by stas00** [comment]
I'm talking about peak memory usage.

if you run torch memory profiler around the reduction call you will see that it'll use `reduce_bucket_size * 4 bytes * 2 copies` - I think it's because one goes from 2b to 4b (1 copy) then torch/nccl makes another copy?

I measured the default `reduce_bucket_size` 5e8 to consume 4GB peak memory usage when comms are in fp32. and only 1GB in bf16.

since those reduction ops memory peaks happen during backward they often lead to OOM if you want to push bs/seqlen.

File: deepspeed/runtime/constants.py (hunk 1 above)
Comment (tohtana) [reply deepspeed/runtime/constants.py:217]: Thank you @stas, fixed to include the other config item.

**Comment by sfc-gh-sbekman** [comment]
Thank you for looking into it,  Masahiro. No problem doing it elsewhere.

Using torch mem profiler will be very helpful to see the reduction memory spikes

https://pytorch.org/blog/understanding-gpu-memory-1/ - it's very easy to set up - if you need help please let me know.

**Comment by sfc-gh-sbekman** [comment]
> Good, I can an assertion to detect that torch.autocast is enabled outside of DeepSpeed but ds_config doesn't set torch_autocast's enabled. Or it might be better to automatically enable it.

If it has to be on and it breaks nothing then automatically enabling it is probably a better idea to help with ease of use.

**Comment by tohtana** [comment]
Hi @stas, 
Thank you for your feedback!

> If it has to be on and it breaks nothing then automatically enabling it is probably a better idea to help with ease of use.

After reviewing the design, I now feel automatically enabling it wouldn't be straightforward. This autocast feature sets some flags to parameters before the optimizer is initialized. However, we only know whether `torch.autocast` is enabled or not just before a forward pass call as `with torch.autocast(...)` is placed to wrap a forward call. Reinitializing parts of the optimizer at that point would complicate the code.

Given that, I think it’s better to [throw an error with the explanation](https://github.com/deepspeedai/DeepSpeed/blob/bb855b762300146f5549bbe98f027438d14c1894/deepspeed/runtime/torch_autocast.py#L97).

**Comment by stas00** [comment]
Then assert is the way to go, Masahiro

**Comment by tohtana** [comment]
> Then assert is the way to go, Masahiro

Thank you @stas00, then can you approve this PR?

**Comment by stas00** [comment]
Hmm, I can't just hit approve, that would be defeat the purpose of doing the review.

We have only discussed one small aspect of this PR, which has been resolved, but the rest of the PR I don't know and currently rushing to finish the porting of Ulysses so until that is done I won't have time to do a serious review.

**Review by tjruwase — APPROVED** [review]

**Comment by Vervious** [comment]
I'm running into an issue where turning on this feature results in massive grad norms, using zero 2; have you seen this before?

```
[deepspeed.torch_autocast]
enabled = true  # NOTE: turning this on makes grad norms explode in stage 2 (but not stage 3)
dtype = "bfloat16"

# [deepspeed.bf16]
# enabled = true  # this works properly

[deepspeed.zero_optimization]
stage = 2
allgather_partitions = true
overlap_comm = false
reduce_scatter = true
contiguous_gradients = true
stage3_prefetch_bucket_size = 0
stage3_max_live_parameters = 0
stage3_max_reuse_distance = 0
stage3_gather_16bit_weights_on_model_save = true

```

Grad norms were reported via `model_engine.get_global_grad_norm()` and also observed via `safe_get_full_grad` after a backward call (but before step). Stage 3 seems to have reasonable grad norms. For some reason the loss curves also don't match exactly between stage 2 and 3 (but they match exactly using deepspeed.bf16 instead)
